### PR TITLE
fix heap-buffer-overflow

### PR DESCRIPTION
**Pull request by zhangtaoXT5**
fix heap-buffer-overflow in https://github.com/redis/hiredis/issues/956

### CODE REVIEW

File: hiredis.c [hunk 1]
```diff
@@ -477,6 +477,8 @@ int redisvFormatCommand(char **target, const char *format, va_list ap) {
 
             touched = 1;
             c++;
+	    if (*c == '\0')
```
Comment (kristjanvalur) [inline hiredis.c:480]: wrong indentation

**Comment by michael-grunder** [comment]
Hi,

I'm going through PRs in preparation for a new release.  I see what you're fixing here, but can you provide an example format string where we would currently read past the end of the buffer?

I thing a regression test may be worthwhile.

File: hiredis.c [hunk 2]
```diff
@@ -477,6 +477,8 @@ int redisvFormatCommand(char **target, const char *format, va_list ap) {
...
+                 continue;
```
Comment (kristjanvalur) [inline hiredis.c:481]:
```suggestion
             continue;
```

**Comment by bjosv** [comment]
I ran the fuzzing tests described in the issue to be able to propose a testcase.
I have pinpointed the problem to the format specifier parser and will push an alternative fix for this, together with the testcases.

**Comment by michael-grunder** [comment]
Closing in favor of #1097.  Thanks for finding and reporting the issue though!

**Comment by kristjanvalur** [comment]
I contest closing this.  Incrementing a zero terminated string twice, without checking that there is indeed a zero in between increments, is dangerous.  One should always check for zeroes, even if one strongly believes that the string is correctly constructed.

File: hiredis.c [hunk 3]
```diff
@@ -477,6 +477,8 @@ int redisvFormatCommand(char **target, const char *format, va_list ap) {
 
             touched = 1;
             c++;
+	    if (*c == '\0')
+             continue;
```
Comment (kristjanvalur) [inline hiredis.c:481]: I'd change this to `break` to make the intention more explicit.

**Comment by gzur** [comment]
I am a professional programmer and I approve of this message.

**Comment by michael-grunder** [comment]
@kristjanvalur I see your point.  It certainly doesn't hurt anything to apply both fixes.

File: hiredis.c (hunk 3 above)
Comment (michael-grunder) [inline hiredis.c:481]:
```suggestion
	    if (*c == '\0')
                 break;
```

File: hiredis.c [hunk 4]
```diff
@@ -477,6 +477,8 @@ int redisvFormatCommand(char **target, const char *format, va_list ap) {
 
             touched = 1;
             c++;
+	    if (*c == '\0')
+                 break;
```
Comment (michael-grunder) [inline hiredis.c:481]:
```suggestion
        if (*c == '\0')
            break;
```

File: hiredis.c [hunk 5]
```diff
@@ -477,6 +477,8 @@ int redisvFormatCommand(char **target, const char *format, va_list ap) {
 
             touched = 1;
             c++;
+        if (*c == '\0')
+            break;
```
Comment (michael-grunder) [inline hiredis.c:481]:
```suggestion
            if (*c == '\0')
                break;
```

**Pull request merged**